### ISSUE TITLE
tools/train.py -c ./configs/det/det_mv3_db.yml   eval model中断

### ISSUE THREAD

**Issue by Cyuliang** [PaddlePaddle/PaddleOCR#2572]
cuda10.2 cudnn7.6
配置文件
Global:
  use_gpu: true
  epoch_num: 500
  log_smooth_window: 20
  print_batch_step: 2
  save_model_dir: ./output/db_mv3/
  save_epoch_step: 100
  # evaluation is run every 2000 iterations
  eval_batch_step: [2000, 1000]
  # if pretrained_model is saved in static mode, load_static_weights must set to True
  load_static_weights: True
  cal_metric_during_train: False
  pretrained_model: C:\Users\cc\Documents\Python\paddlepaddle-PaddleOCR-release-2.0\PaddleOCR\model\MobileNetV3_large_x0_5_pretrained\
  checkpoints: C:\Users\cc\Documents\Python\paddlepaddle-PaddleOCR-release-2.0\PaddleOCR\output\db_mv3\latest
  save_inference_dir:
  use_visualdl: False
  infer_img: doc/imgs_en/img_10.jpg
  save_res_path: ./output/det_db/predicts_db.txt
  character_dict_path: C:\Users\cc\Documents\Python\paddlepaddle-PaddleOCR-release-2.0\PaddleOCR\predict.txt
  max_text_length: 11
  character_type: en

训练
[2021/04/20 20:41:01] root INFO: epoch: [20/500], iter: 674, lr: 0.001000, loss: 1.868258, loss_shrink_maps: 1.050074, loss_threshold_maps: 0.613446, loss_binary_maps: 0.210883, reader_cost: 0.08675 s, batch_cost: 0.26231 s, samples: 8, ips: 15.24896
[2021/04/20 20:41:02] root INFO: epoch: [20/500], iter: 676, lr: 0.001000, loss: 1.766201, loss_shrink_maps: 0.944057, loss_threshold_maps: 0.613446, loss_binary_maps: 0.187976, reader_cost: 0.07433 s, batch_cost: 0.25932 s, samples: 8, ips: 15.42496
[2021/04/20 20:41:03] root INFO: epoch: [20/500], iter: 678, lr: 0.001000, loss: 1.750933, loss_shrink_maps: 0.944057, loss_threshold_maps: 0.613446, loss_binary_maps: 0.187976, reader_cost: 0.05586 s, batch_cost: 0.23141 s, samples: 8, ips: 17.28509
[2021/04/20 20:41:04] root INFO: epoch: [20/500], iter: 680, lr: 0.001000, loss: 1.737311, loss_shrink_maps: 0.912175, loss_threshold_maps: 0.596961, loss_binary_maps: 0.181780, reader_cost: 0.09723 s, batch_cost: 0.27878 s, samples: 8, ips: 14.34848
[2021/04/20 20:41:05] root INFO: epoch: [20/500], iter: 682, lr: 0.001000, loss: 1.750933, loss_shrink_maps: 0.938082, loss_threshold_maps: 0.596961, loss_binary_maps: 0.186815, reader_cost: 0.06980 s, batch_cost: 0.25533 s, samples: 8, ips: 15.66591
[2021/04/20 20:41:06] root INFO: epoch: [20/500], iter: 684, lr: 0.001000, loss: 1.750933, loss_shrink_maps: 0.938082, loss_threshold_maps: 0.599591, loss_binary_maps: 0.186815, reader_cost: 0.08327 s, batch_cost: 0.27579 s, samples: 8, ips: 14.50374
[2021/04/20 20:41:07] root INFO: epoch: [20/500], iter: 686, lr: 0.001000, loss: 1.737311, loss_shrink_maps: 0.894709, loss_threshold_maps: 0.588326, loss_binary_maps: 0.178650, reader_cost: 0.08079 s, batch_cost: 0.26281 s, samples: 8, ips: 15.21994
[2021/04/20 20:41:07] root INFO: epoch: [20/500], iter: 688, lr: 0.001000, loss: 1.737311, loss_shrink_maps: 0.901397, loss_threshold_maps: 0.593149, loss_binary_maps: 0.179552, reader_cost: 0.09275 s, batch_cost: 0.27229 s, samples: 8, ips: 14.69048
[2021/04/20 20:41:08] root INFO: epoch: [20/500], iter: 690, lr: 0.001000, loss: 1.737311, loss_shrink_maps: 0.901397, loss_threshold_maps: 0.588326, loss_binary_maps: 0.179552, reader_cost: 0.07928 s, batch_cost: 0.25930 s, samples: 8, ips: 15.42602
[2021/04/20 20:41:09] root INFO: epoch: [20/500], iter: 692, lr: 0.001000, loss: 1.737311, loss_shrink_maps: 0.901397, loss_threshold_maps: 0.588326, loss_binary_maps: 0.179552, reader_cost: 0.06232 s, batch_cost: 0.24886 s, samples: 8, ips: 16.07326
[2021/04/20 20:41:10] root INFO: epoch: [20/500], iter: 694, lr: 0.001000, loss: 1.786281, loss_shrink_maps: 1.006736, loss_threshold_maps: 0.593149, loss_binary_maps: 0.200538, reader_cost: 0.08987 s, batch_cost: 0.26878 s, samples: 8, ips: 14.88211

eval model:: 96%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▌ | 49/51 [00:02<00:00, 25.03it/s]

**Comment by tink2123** [comment]
单独eval，检查下数据有没有问题。 另外减小batch_size 和 num_worker 试下

**Comment by Jane-Ding** [comment]
修改了tools/program.py 里的eval 方法，计数器 idx 是从0开始的，如果一共10张图片，最大idx=9，改了条件判断并放在最后：
if idx >= len(valid_dataloader)-1:
break
可以运行了：
def eval(model, valid_dataloader, post_process_class, eval_class,
use_srn=False):
model.eval()
with paddle.no_grad():
total_frame = 0.0
total_time = 0.0
pbar = tqdm(total=len(valid_dataloader), desc='eval model:')
for idx, batch in enumerate(valid_dataloader):
# if idx >= len(valid_dataloader):
# break
images = batch[0]
start = time.time()
if use_srn:
others = batch[-4:]
preds = model(images, others)
else:
preds = model(images)
batch = [item.numpy() for item in batch]
# Obtain usable results from post-processing methods
post_result = post_process_class(preds, batch[1])
total_time += time.time() - start
# Evaluate the results of the current batch
eval_class(post_result, batch)
pbar.update(1)
total_frame += len(images)
# change idx and move to here
if idx >= len(valid_dataloader)-1:
break
# Get final metric，eg. acc or hmean
metric = eval_class.get_metric()
print('get metric: ', metric)
pbar.close()
model.train()
metric['fps'] = total_frame / total_time
return metric

**Comment by yfc886** [comment]
> 修改了tools/program.py 里的eval 方法，计数器 idx 是从0开始的，如果一共10张图片，最大idx=9，改了条件判断并放在最后：
> if idx >= len(valid_dataloader)-1:
> break
> 可以运行了：
> def eval(model, valid_dataloader, post_process_class, eval_class,
> use_srn=False):
> model.eval()
> with paddle.no_grad():
> total_frame = 0.0
> total_time = 0.0
> pbar = tqdm(total=len(valid_dataloader), desc='eval model:')
> for idx, batch in enumerate(valid_dataloader):
> 
> # if idx >= len(valid_dataloader):
> # break
> images = batch[0]
> start = time.time()
> if use_srn:
> others = batch[-4:]
> preds = model(images, others)
> else:
> preds = model(images)
> batch = [item.numpy() for item in batch]
> 
> # Obtain usable results from post-processing methods
> post_result = post_process_class(preds, batch[1])
> total_time += time.time() - start
> 
> # Evaluate the results of the current batch
> eval_class(post_result, batch)
> pbar.update(1)
> total_frame += len(images)
> 
> # change idx and move to here
> if idx >= len(valid_dataloader)-1:
> break
> 
> # Get final metric，eg. acc or hmean
> metric = eval_class.get_metric()
> print('get metric: ', metric)
> pbar.close()
> model.train()
> metric['fps'] = total_frame / total_time
> return metric

感谢大佬，确实有效。

**Comment by Jane-Ding** [comment]
> > 修改了tools/program.py 里的eval 方法，计数器 idx 是从0开始的，如果一共10张图片，最大idx=9，改了条件判断并放在最后：
> > if idx >= len(valid_dataloader)-1:
> > break
> > 可以运行了：
> > def eval(model, valid_dataloader, post_process_class, eval_class,
> > use_srn=False):
> > model.eval()
> > with paddle.no_grad():
> > total_frame = 0.0
> > total_time = 0.0
> > pbar = tqdm(total=len(valid_dataloader), desc='eval model:')
> > for idx, batch in enumerate(valid_dataloader):
> > # if idx >= len(valid_dataloader):
> > # break
> > images = batch[0]
> > start = time.time()
> > if use_srn:
> > others = batch[-4:]
> > preds = model(images, others)
> > else:
> > preds = model(images)
> > batch = [item.numpy() for item in batch]
> > # Obtain usable results from post-processing methods
> > post_result = post_process_class(preds, batch[1])
> > total_time += time.time() - start
> > # Evaluate the results of the current batch
> > eval_class(post_result, batch)
> > pbar.update(1)
> > total_frame += len(images)
> > # change idx and move to here
> > if idx >= len(valid_dataloader)-1:
> > break
> > # Get final metric，eg. acc or hmean
> > metric = eval_class.get_metric()
> > print('get metric: ', metric)
> > pbar.close()
> > model.train()
> > metric['fps'] = total_frame / total_time
> > return metric
> 
> 感谢大佬，确实有效。

并且把tools/program.py 里的train 方法，也改一下计数器，大概在201行，if idx >= len(train_dataloader)-1:
break， 我在win10机器上就能正常训练，如果显存不大的话，把batch size调小，worker num调成1. 很奇怪的是，我在colab上都能正常训练。

**Comment by sml2h3** [comment]
找到个最靠谱的解决方法，不用改代码。
下载一个everything软件，然后搜索msvcp140.dll，按大小排序，把最大的那个msvcp140.dll替换掉C:\Windows\System32中的msvcp140.dll再启动项目就可以了，亲测可用
